File: sdk/gsusb-cpp/examples/02_periodic.cpp
#include "demo_common.hpp"

#ifndef WIN32_LEAN_AND_MEAN
#define WIN32_LEAN_AND_MEAN
#endif

#ifndef NOMINMAX
#define NOMINMAX
#endif

#include <cmath>
#include <iostream>
#include <windows.h>

int main(int argc, char** argv) {
    try {
        const auto options = demo::parse_common_options(argc, argv, "Single-channel periodic send example");
        auto tx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.tx_channel, false));
        auto rx_bus = std::make_unique<gsusb::GsUsbBus>(demo::make_bus_options(options, options.rx_channel, false));

        const int count = std::max(1, static_cast<int>(std::lround(options.duration_s / std::max(options.period_s, 0.001))));
        std::uint64_t tx_count = 0;
        std::uint64_t rx_count = 0;
        for (int index = 0; index < count; ++index) {
            gsusb::CanMessage message;
            message.arbitration_id = 0x100;
            message.data = {static_cast<std::uint8_t>(index & 0xFF), 0x01, 0x02, 0x03};
            tx_bus->send(message);
            ++tx_count;
            if (auto received = rx_bus->recv(std::chrono::milliseconds(static_cast<int>(options.rx_timeout_s * 1000.0))); received.has_value()) {
                ++rx_count;
                demo::print_message("Received", received.value());
            }
            Sleep(static_cast<DWORD>(std::max(0.0, options.period_s * 1000.0)));
        }

        std::cout << "Periodic summary TX=" << tx_count << " RX=" << rx_count << "\n";
        tx_bus->shutdown();
        rx_bus->shutdown();
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << ex.what() << "\n";
        return 1;
    }
}

File: sdk/gsusb-rs/examples/02_periodic.rs
#[path = "common/mod.rs"]
mod common;

use common::{make_bus_options, parse_common_options, print_message};
use gsusb_rs::{CanMessage, GsUsbBus, Result};
use std::env;
use std::time::Duration;

fn main() -> Result<()> {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let options = parse_common_options(&args, "Single-channel periodic send example")?;
    let tx_bus = GsUsbBus::new(make_bus_options(&options, options.tx_channel, false, None, None))?;
    let rx_bus = GsUsbBus::new(make_bus_options(&options, options.rx_channel, false, None, None))?;

    let count = ((options.duration_s / options.period_s.max(0.001)).round() as i32).max(1);
    let mut tx_count = 0u64;
    let mut rx_count = 0u64;
    for index in 0..count {
        let message = CanMessage {
            arbitration_id: 0x100,
            data: vec![index as u8, 0x01, 0x02, 0x03],
            ..CanMessage::default()
        };
        tx_bus.send(&message)?;
        tx_count += 1;
        if let Some(received) = rx_bus.recv(Duration::from_secs_f64(options.rx_timeout_s)) {
            rx_count += 1;
            print_message("Received", &received);
        }
        std::thread::sleep(Duration::from_secs_f64(options.period_s.max(0.0)));
    }

    println!("Periodic summary TX={tx_count} RX={rx_count}");
    tx_bus.shutdown();
    rx_bus.shutdown();
    Ok(())
}